core/services/ccip/orm.go

```diff
@@ -79,82 +71,143 @@ func (o *orm) GetGasPricesByDestChain(ctx context.Context, destChainSelector uin
+func (o *orm) UpsertTokenPricesForDestChain(ctx context.Context, destChainSelector uint64, tokenPrices []TokenPrice, interval time.Duration) (int64, error) {
 	if len(tokenPrices) == 0 {
-		return nil
+		return 0, nil
+	}
+
+	tokensToUpdate, err := o.pickOnlyRelevantTokensForUpdate(ctx, destChainSelector, tokenPrices, interval)
+	if err != nil || len(tokensToUpdate) == 0 {
+		return 0, err
 	}
 
-	insertData := make([]map[string]interface{}, 0, len(tokenPrices))
-	for _, price := range tokenPrices {
+	insertData := make([]map[string]interface{}, 0, len(tokensToUpdate))
+	for _, price := range tokensToUpdate {
 		insertData = append(insertData, map[string]interface{}{
 			"chain_selector": destChainSelector,
-			"job_id":         jobId,
 			"token_addr":     price.TokenAddr,
 			"token_price":    price.TokenPrice,
 		})
 	}
 
-	// using statement_timestamp() to make testing easier
-	stmt := `INSERT INTO ccip.observed_token_prices (chain_selector, job_id, token_addr, token_price, created_at)
-		VALUES (:chain_selector, :job_id, :token_addr, :token_price, statement_timestamp());`
-	_, err := o.ds.NamedExecContext(ctx, stmt, insertData)
+	stmt := `INSERT INTO ccip.observed_token_prices (chain_selector, token_addr, token_price, created_at)
+		VALUES (:chain_selector, :token_addr, :token_price, statement_timestamp())
+		ON CONFLICT (token_addr, chain_selector) 
+		DO UPDATE SET token_price = EXCLUDED.token_price, created_at = EXCLUDED.created_at;`
+	result, err := o.ds.NamedExecContext(ctx, stmt, insertData)
 	if err != nil {
-		err = fmt.Errorf("error inserting token prices for job %d: %w", jobId, err)
+		return 0, fmt.Errorf("error inserting token prices %w", err)
 	}
-
-	return err
+	return result.RowsAffected()
 }
 
-func (o *orm) ClearGasPricesByDestChain(ctx context.Context, destChainSelector uint64, expireSec int) error {
-	stmt := `DELETE FROM ccip.observed_gas_prices WHERE chain_selector = $1 AND created_at < (statement_timestamp() - $2 * interval '1 second')`
+// pickOnlyRelevantTokensForUpdate returns only tokens that need to be updated. Multiple jobs can be updating the same tokens,
+// in order to reduce table locking and redundant upserts we start with reading the table and checking which tokens are eligible for update.
+// A token is eligible for update when:
+// * price has changed and created_at date is older than the interval
+// * it's not present in the result set from the db query (e.g. it's a new token)
+// Therefore if there are no price changes for a single token we won't update the state of the database
+func (o *orm) pickOnlyRelevantTokensForUpdate(
+	ctx context.Context,
+	destChainSelector uint64,
+	tokenPrices []TokenPrice,
+	interval time.Duration,
+) ([]TokenPrice, error) {
+	stmt := `
+		SELECT token_addr, token_price, created_at
+		FROM ccip.observed_token_prices
+		WHERE chain_selector = $1 and token_addr = ANY($2);
```

Comment:
Might be cheap to have an index on `created_at` and add the condition and `created_at < $3`. Not a blocker.